pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java

```diff
@@ -259,6 +271,33 @@ public void channelRead(final ChannelHandlerContext ctx, Object msg) throws Exce
         }
     }
 
+    /**
+     * Use splice to zero-copy of NIC to NIC.
+     * @param inboundChannel input channel
+     * @param outboundChannel output channel
+     */
+    protected void spliceNIC2NIC(EpollSocketChannel inboundChannel, EpollSocketChannel outboundChannel) {
+        inboundChannel.config().setAutoRead(false);
+
+        ChannelPromise promise = ctx.newPromise();
+        inboundChannel.spliceTo(outboundChannel, SPLICE_BYTES, promise);
+        promise.addListener((ChannelFutureListener) future -> {
+            if (!future.isSuccess()) {
+                future.channel().pipeline().fireExceptionCaught(future.cause());
+            } else {
+                ProxyService.OPS_COUNTER.inc();
+                directProxyHandler.getInboundChannelRequestsRate().recordEvent(SPLICE_BYTES);
+                ProxyService.BYTES_COUNTER.inc(SPLICE_BYTES);
+            }
+        });
+
+        inboundChannel.config().setAutoRead(true);
```

Comment:
Doesn't this have to happen in the promise listener after the `spliceTo` operation has completed?

Reply:
Thanks, I remove the settings of `autoRead`.